browser/ui/views/brave_actions/brave_shields_action_view.cc

```diff
@@ -0,0 +1,131 @@
+    : LabelButton(base::BindRepeating(&BraveShieldsActionView::ButtonPressed,
+                                      base::Unretained(this)),
+                  std::u16string()),
+      delegate_(delegate) {
+  auto* ink_drop = views::InkDrop::Get(this);
+  ink_drop->SetMode(views::InkDropHost::InkDropMode::ON);
+  ink_drop->SetBaseColorCallback(base::BindRepeating(
+      [](views::View* host) { return GetToolbarInkDropBaseColor(host); },
+      this));
+
+  SetAccessibleName(l10n_util::GetStringUTF16(IDS_BRAVE_UI_BRAVE_REWARDS));
+  SetHasInkDropActionOnClick(true);
+  SetHorizontalAlignment(gfx::ALIGN_CENTER);
+  ink_drop->SetVisibleOpacity(kToolbarInkDropVisibleOpacity);
+  // Create badge-and-image source like an extension icon would
+  auto preferred_size = GetPreferredSize();
+  ui::ResourceBundle& rb = ui::ResourceBundle::GetSharedInstance();
+  std::unique_ptr<IconWithBadgeImageSource> image_source(
+      new BraveActionIconWithBadgeImageSource(preferred_size));
+  // Set icon on badge using actual extension icon resource
+  gfx::ImageSkia image;
+
+  const SkBitmap bitmap =
+      rb.GetImageNamed(IDR_BRAVE_SHIELDS_ICON_64).AsBitmap();
+  float scale = static_cast<float>(bitmap.width()) / kBraveActionGraphicSize;
+  image.AddRepresentation(gfx::ImageSkiaRep(bitmap, scale));
+  image_source->SetIcon(gfx::Image(image));
+  // TODO(nullhook): Create badge and set text on badge via pref text
+  gfx::ImageSkia icon(
+      gfx::Image(gfx::ImageSkia(std::move(image_source), preferred_size))
+          .AsImageSkia());
+  // Use badge-and-icon source for button's image in all states
+  SetImage(views::Button::STATE_NORMAL, icon);
+  // Install highlight path generator
+  views::HighlightPathGenerator::Install(
+      this, std::make_unique<BraveShieldsActionViewHighlightPathGenerator>());
+}
+
+BraveShieldsActionView::~BraveShieldsActionView() {}
+
+SkPath BraveShieldsActionView::GetHighlightPath() const {
+  // Set the highlight path for the toolbar button,
+  // making it inset so that the badge can show outside it in the
+  // fake margin on the right that we are creating.
+  gfx::Insets highlight_insets(0, 0, 0, kBraveActionRightMargin);
+  gfx::Rect rect(GetPreferredSize());
+  rect.Inset(highlight_insets);
+  const int radii = ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
+      views::Emphasis::kMaximum, rect.size());
+  SkPath path;
+  path.addRoundRect(gfx::RectToSkRect(rect), radii, radii);
+  return path;
+}
+
+void BraveShieldsActionView::ButtonPressed() {
+  NOTIMPLEMENTED();
+}
+
+gfx::Size BraveShieldsActionView::CalculatePreferredSize() const {
+  return delegate_->GetToolbarActionSize();
+}
+
+std::unique_ptr<views::LabelButtonBorder>
+BraveShieldsActionView::CreateDefaultBorder() const {
+  std::unique_ptr<views::LabelButtonBorder> border =
+      LabelButton::CreateDefaultBorder();
+  border->set_insets(gfx::Insets(0, 0, 0, 0));
+  return border;
+}
+
+void BraveShieldsActionView::Update() {
```

Comment:
This gets called by BraveActionsContainer::Update, which will probably get called when the active tab is changed which is needed so that we can update with shields count and enabled/disabled state for that new active tab. But we're also going to want to call `Update` when the shields blocked count changes, and when shields is enabled / disabled on the current tab. Just a TODO that we'll need to wire up some kind of layer-appropriate observer.